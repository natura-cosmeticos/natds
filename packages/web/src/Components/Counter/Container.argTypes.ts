import { ArgType } from "@storybook/addons";
import { ICounterProps } from "./Counter.props";

export const argTypes : Record<keyof ICounterProps | string, ArgType> = {

  label: {
    defaultValue: null,
    type: "string",
  },
  readOnly: {
    defaultValue: false,
    type: "boolean",
  },
  size: {
    defaultValue: "small",
    type: "string",
  },
  initialValue: {
    defaultValue: 1,
    type: "number",
  },
  maxValue: {
    defaultValue: 99,
    type: "number",
  },
  minValue: {
    defaultValue: 0,
    type: "number",
  },
  onChange: {
    defaultValue: null,
    type: "function",
  },
  onIncrement: {
    defaultValue: null,
    type: "function",
  },
  onDecrement: {
    defaultValue: null,
    type: "function",
  },
};

export default argTypes;
