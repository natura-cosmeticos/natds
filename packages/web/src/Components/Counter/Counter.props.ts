export type Sizes = "small" | "medium";

export interface ICounterProps {
  label?: string,
  readOnly?: boolean
  size?: Sizes,
  initialValue: number,
  maxValue?: number,
  minValue?: number,
  onChange: (newValue: number) => void,
  onIncrement: (newValue: number) => void,
  onDecrement: (newValue: number) => void,
}
