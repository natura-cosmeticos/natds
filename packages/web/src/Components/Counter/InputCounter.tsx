/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
import * as React from 'react'
import { ITextFieldProps, TextField } from '../TextField'

export { ICounterProps } from './Counter.props'

export const InputCounter = ({ value, className, onChange }: Pick<ITextFieldProps, 'value' | 'className' | 'onChange'>) => (
  <TextField
    color="secondary"
    value={value}
    className={className}
    type="number"
    onChange={onChange}
  />
)

export default InputCounter
