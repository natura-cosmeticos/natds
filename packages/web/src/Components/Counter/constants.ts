export const QUANTITY = 1
export const INPUT_DIFFERENCE = 2
export const MAX_VALUE = 99
export const MIN_VALUE = 0
