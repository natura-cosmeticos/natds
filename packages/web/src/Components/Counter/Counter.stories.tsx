import * as React from 'react'
import { Meta, Story } from '@storybook/react'
import { Counter as component, ICounterProps } from './Counter'
import argTypes from './Container.argTypes'

const Counter : React.ForwardRefExoticComponent<ICounterProps> = React.lazy(() => import('./Counter'))

export default {
  argTypes,
  component,
  title: 'Components/Counter'
} as Meta

const Template : Story<ICounterProps> = (args: ICounterProps) => (
  <div>
    <Counter {...args} />
  </div>
)

export const Playground : Story<ICounterProps> = Template.bind({})
Playground.args = {
  onChange: () => null,
  onDecrement: () => null,
  onIncrement: () => null
}

export const SmallSize : Story<ICounterProps> = Template.bind({})
SmallSize.args = {
  ...Playground.args,
  size: 'small'
}

export const MediumSize : Story<ICounterProps> = Template.bind({})
MediumSize.args = {
  ...Playground.args,
  size: 'medium'
}

export const WithLabel : Story<ICounterProps> = Template.bind({})
WithLabel.args = {
  ...Playground.args,
  size: 'small',
  label: 'Quantidade'
}
