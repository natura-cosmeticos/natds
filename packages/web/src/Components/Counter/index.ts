import Counter from './Counter'

export { ICounterProps } from './Counter.props'
export { Counter }
export default Counter
