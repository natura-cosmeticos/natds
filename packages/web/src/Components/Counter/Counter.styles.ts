import createStyles from "@material-ui/core/styles/createStyles";
import makeStyles from "@material-ui/core/styles/makeStyles";
import { tokens } from "@naturacosmeticos/natds-styles";
import { ICounterProps } from "./Counter";
import { IThemeWeb } from "../../Themes";
import { INPUT_DIFFERENCE } from "./constants";

const { sizes } = tokens;

const style = (theme: IThemeWeb) => createStyles({
  button: {
    fontSize: 14,
    minHeight: ({ size }: ICounterProps) => `${size === "small" ? theme.sizes?.semiX : theme.sizes?.medium}px`,
    "&:last-child": {
      borderLeft: 0,
    },
  },
  input: {
    textAlign: "center",
    width: `${sizes.semiX}px`,
    "& .MuiInput-input": {
      padding: 0,
      borderRadius: 0,
      margin: 1,
      minHeight: ({ size }: ICounterProps) => `${(size === "small" ? sizes?.semiX : sizes?.medium) - INPUT_DIFFERENCE}px`,
      fontSize: theme.typography.body1?.fontSize,
      "&::-webkit-inner-spin-button, &::-webkit-outer-spin-button": {
        WebkitAppearance: "none",
      },
    },
  },
});

export const useStyles = makeStyles(style, { name: "NatDSCounter" });

export default useStyles;
