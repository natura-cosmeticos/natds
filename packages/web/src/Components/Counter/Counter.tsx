/* eslint-disable no-unused-expressions */
import * as React from "react";
import { ICounterProps } from "./Counter.props";
import { TextField } from "../TextField";
import { useStyles } from "./Counter.styles";
import { ButtonGroup } from "../ButtonGroup";
import { Button } from "../Button";
import { InputLabel } from "../InputLabel";
import { QUANTITY, MAX_VALUE, MIN_VALUE } from "./constants";

export { ICounterProps } from "./Counter.props";

/**
 * ## Importing
 *
 * ```
 * import { Counter } from '@naturacosmeticos/natds-web';
 * ```
 */

export const Counter = React.forwardRef<unknown, ICounterProps>((props: ICounterProps) => {

  const {
    onChange,
    onDecrement,
    onIncrement,
    size,
    label,
    readOnly,
    initialValue,
    maxValue = MAX_VALUE,
    minValue = MIN_VALUE,
  } = props;

  const { button, input } = useStyles(props);

  const [value, setValue] = React.useState(initialValue || QUANTITY);

  // eslint-disable-next-line consistent-return
  React.useEffect(() => {
    if (initialValue > MAX_VALUE) {
      return setValue(MAX_VALUE);
    }
    if (initialValue < MIN_VALUE) {
      return setValue(MIN_VALUE);
    }
  }, [initialValue]);

  const handleInputChange = (event: { target: HTMLInputElement }) => {
    const newValue = Number(event.target.value);

    if (!newValue || (newValue >= minValue && newValue <= maxValue)) {
      setValue(newValue);
      onChange && onChange(newValue);
    }
  };

  const increment = () => {
    const newValue = value + QUANTITY;

    if (newValue <= maxValue) {
      setValue(newValue);
      onIncrement && onIncrement(newValue);
    }
  };

  const decrement = () => {
    const newValue = value - QUANTITY;

    if (newValue >= minValue) {
      setValue(newValue);
      onDecrement && onDecrement(newValue);
    }
  };

  const maxReached = value >= maxValue;
  const minReached = value <= minValue;

  return (
    <div>
      { label && <InputLabel>{ label }</InputLabel> }
      <ButtonGroup
        color="default"
        disableElevation
        size={size}
        orientation="horizontal"
        disabled={readOnly}
      >
        <Button
          disabled={minReached}
          variant="outlined"
          onClick={decrement}
          className={button}>
          -
        </Button>
        <TextField
          value={value}
          className={input}
          type="number"
          onChange={handleInputChange}
        />
        <Button
          disabled={maxReached}
          variant="outlined"
          onClick={increment}
          className={button}>
          +
        </Button>
      </ButtonGroup>
    </div>
  );
});

Counter.displayName = "Counter";

Counter.defaultProps = {
  readOnly: false,
  size: "small",
  minValue: MIN_VALUE,
  maxValue: MAX_VALUE,
};

export default Counter;
