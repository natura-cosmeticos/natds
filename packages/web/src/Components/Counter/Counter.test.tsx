import * as React from "react";
import * as TestRenderer from "react-test-renderer";

import Counter from "./Counter";
import { InputLabel } from "../InputLabel";
import { TextField } from "../TextField";

describe("Counter component", () => {

  it("renders correctly", () => {
    const testRenderer : TestRenderer.ReactTestRenderer = TestRenderer.create(
      <Counter
        initialValue={1}
        onChange={jest.fn()}
        onIncrement={jest.fn()}
        onDecrement={jest.fn()}
      />,
    );

    expect(testRenderer).toMatchSnapshot("Counter component renders correctly");
  });

  it("renders with label", () => {
    const labelValue = "Contador";
    const testRenderer : TestRenderer.ReactTestRenderer = TestRenderer.create(
      <Counter
        initialValue={1}
        label={labelValue}
        onChange={jest.fn()}
        onIncrement={jest.fn()}
        onDecrement={jest.fn()}
      />,
    );

    expect(testRenderer).toMatchSnapshot("Counter component renders with label");
    const testInstance = testRenderer.root;

    expect(testInstance.findByType(InputLabel).props.children).toEqual(labelValue);
  });

  it("should set the correct value when initial is invalid", () => {
    const testRenderer : TestRenderer.ReactTestRenderer = TestRenderer.create(
      <Counter
        initialValue={200}
        onChange={jest.fn()}
        onIncrement={jest.fn()}
        onDecrement={jest.fn()}
      />,
    );

    const testRenderer2 : TestRenderer.ReactTestRenderer = TestRenderer.create(
      <Counter
        initialValue={-1}
        onChange={jest.fn()}
        onIncrement={jest.fn()}
        onDecrement={jest.fn()}
      />,
    );

    const testInstance = testRenderer.root;
    const testInstance2 = testRenderer2.root;

    expect(testInstance.findByType(TextField).props.value).toEqual(testInstance.props.maxValue);
    expect(testInstance2.findByType(TextField).props.value).toEqual(testInstance.props.minValue);
  });
});
