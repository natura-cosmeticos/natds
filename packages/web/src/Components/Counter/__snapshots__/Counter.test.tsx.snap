// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component renders correctly: Counter component renders correctly 1`] = `
<div>
  <div
    className="MuiButtonGroup-root MuiButtonGroup-disableElevation"
    role="group"
  >
    <button
      className="MuiButtonBase-root MuiButton-root NatDSButton-root-5 MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined NatDSCounter-button-1 NatDSCounter-button-3 MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label NatDSButton-label-6"
      >
        -
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <div
      className="MuiFormControl-root NatDSFormControl-root-8 MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined NatDSCounter-input-2 NatDSCounter-input-4"
    >
      <div
        className="MuiInputBase-root MuiInput-root NatDSInput-root-14 MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        disableElevation={true}
        disableFocusRipple={false}
        disableRipple={false}
        onClick={[Function]}
        required={false}
        size="small"
        variant="outlined"
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input NatDSInput-input-11 NatDSInput-input-15 MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined NatDSCounter-input-2 NatDSCounter-input-4__input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          required={false}
          type="number"
          value={1}
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root NatDSButton-root-5 MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined NatDSCounter-button-1 NatDSCounter-button-3 MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label NatDSButton-label-6"
      >
        +
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`Counter component renders with label: Counter component renders with label 1`] = `
<div>
  <label
    className="MuiFormLabel-root MuiInputLabel-root NatDSInputLabel-root-18 NatDSInputLabel-root-19 MuiInputLabel-animated MuiInputLabel-shrink"
    data-shrink={true}
  >
    Contador
  </label>
  <div
    className="MuiButtonGroup-root MuiButtonGroup-disableElevation"
    role="group"
  >
    <button
      className="MuiButtonBase-root MuiButton-root NatDSButton-root-5 MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined NatDSCounter-button-1 NatDSCounter-button-16 MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label NatDSButton-label-6"
      >
        -
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
    <div
      className="MuiFormControl-root NatDSFormControl-root-8 MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined NatDSCounter-input-2 NatDSCounter-input-17"
    >
      <div
        className="MuiInputBase-root MuiInput-root NatDSInput-root-14 MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        disableElevation={true}
        disableFocusRipple={false}
        disableRipple={false}
        onClick={[Function]}
        required={false}
        size="small"
        variant="outlined"
      >
        <input
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input NatDSInput-input-11 NatDSInput-input-20 MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined NatDSCounter-input-2 NatDSCounter-input-17__input"
          disabled={false}
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder=""
          required={false}
          type="number"
          value={1}
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiButton-root NatDSButton-root-5 MuiButton-outlined MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedOutlined MuiButtonGroup-groupedOutlinedHorizontal MuiButtonGroup-groupedOutlined NatDSCounter-button-1 NatDSCounter-button-16 MuiButton-outlinedSizeSmall MuiButton-sizeSmall MuiButton-disableElevation"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label NatDSButton-label-6"
      >
        +
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
